src/Streamly/Streams/StreamD.hs

```diff
@@ -2558,25 +2559,43 @@ data CodePoint = FreshPoint
                | GoFour !Word8
                | Four1 !Word8 !Word8
                | Four !Word8 !Word8 !Word8
-               | YieldPoint !Char
-
-{-# INLINE_NORMAL decodeUtf8 #-}
-decodeUtf8 :: Monad m => Stream m Word8 -> Stream m Char
-decodeUtf8 (Stream step state) = Stream step' (state, FreshPoint)
+               | YieldAndContinue !Char
+               | YieldAndStop !Char
+               | Done
+
+data CodingFailureMode
+    = TransliterateCodingFailure
+    | ErrorOnCodingFailure
+    deriving (Show)
+
+{-# INLINE_NORMAL decodeUtf8With #-}
+decodeUtf8With :: Monad m => CodingFailureMode -> Stream m Word8 -> Stream m Char
+decodeUtf8With cfm (Stream step state) = Stream step' (state, FreshPoint)
   where
+    {-# INLINE transliterateOrError #-}
+    {-# INLINE inputUnderflow #-}
+    transliterateOrError e s =
+        case cfm of
+            ErrorOnCodingFailure -> error e
+            TransliterateCodingFailure -> Skip (s, YieldAndContinue '\xFFFD')
+    inputUnderflow s = case cfm of
+       ErrorOnCodingFailure -> error "Input Underflow"
+       TransliterateCodingFailure -> Skip (s, YieldAndStop '\xFFFD')
+
     {-# INLINE_LATE step' #-}
     step' gst (st, FreshPoint) = do
         r <- step (adaptState gst) st
         return $
             case r of
                 Yield c0 s
                     | c0 <= 0x7f ->
-                        Skip (s, YieldPoint $ unsafeChr (fromIntegral c0))
-                    | c0 >= 0xc0 && c0 <= 0xc1 -> error "Invalid Overlong forms"
+                        Skip (s, YieldAndContinue $ unsafeChr (fromIntegral c0))
+                    | c0 >= 0xc0 && c0 <= 0xc1 ->
+                        transliterateOrError "Invalid Overlong forms" s
```

Comment:
All errors should have some identification to tell the programmer/user where they are coming from. Ideally we should also print some context and what exactly is invalid so that the user can debug.